KAFKA-15931: Cancel RemoteLogReader gracefully

jeqo opened the pull request:
Backports f24945b519005c0bc7a28db2db7aae6cec158927 to 3.9

Instead of reopening the transaction index, it cancels the
RemoteFetchTask without interrupting it--avoiding to close the
TransactionIndex channel.

This will lead to complete the execution of the remote fetch but
ignoring
the results. Given that this is considered a rare case, we could live
with this. If it becomes a performance issue, it could be optimized.

junrao commented:
@jeqo : Thanks for the PR. We probably want to hold off on this PR under https://github.com/apache/kafka/pull/15241#discussion_r1988994745 is resolved.

jeqo committed:
KAFKA-15931: Cancel RemoteLogReader gracefully

Backports f24945b519005c0bc7a28db2db7aae6cec158927 to 3.9

Instead of reopening the transaction index, it cancels the
RemoteFetchTask without interrupting it--avoiding to close the
TransactionIndex channel.

This will lead to complete the execution of the remote fetch but
ignoring
the results. Given that this is considered a rare case, we could live
with this. If it becomes a performance issue, it could be optimized.

jeqo commented:
@junrao I've updated the PR with the latest fix. PTAL, many thanks!

junrao reviewed (approved):
@jeqo : Thanks for the PR. The code LGTM. Are the test failures related?

junrao commented:
Also, do you plan to do a cherry-pick for 4.0 too?

jeqo commented:
The test failure seem unrelated, yes. Tested all failing tests locally and they passed:

- TransactionsWithTieredStoreTest
- ConsumerBounceTest
- PurgeRepartitionTopicIntegrationTest 
- CoordinatorTest

> Also, do you plan to do a cherry-pick for 4.0 too?

Good call. I could do that too.

jeqo commented:
Opened https://github.com/apache/kafka/pull/19331 for 4.0 backport

junrao commented:
@jeqo : Thanks for the reply. Merging this PR.
